src/WtsApi32.Desktop/WtsApi32.cs

```diff
@@ -0,0 +1,112 @@
+        /// server, or <see cref="WTS_CURRENT_SERVER_HANDLE"/> to use the RD Session Host server that hosts your application.</param>
+        /// <param name="reserved">This parameter is reserved. It must be zero.</param>
+        /// <param name="version">The version of the enumeration request. This parameter must be 1.</param>
+        /// <param name="ppSessionInfo">A pointer to an array of <see cref="WTS_SESSION_INFO"/> structures that represent the retrieved
+        /// sessions. To free the returned buffer, call the <see cref="WTSFreeMemory"/> function.</param>
+        /// <param name="pSessionInfoCount">A pointer to the number of WTS_SESSION_INFO structures returned in the ppSessionInfo parameter.</param>
+        /// <returns>Returns zero if this function fails. If this function succeeds, a nonzero value is returned.</returns>
+        [DllImport(nameof(WtsApi32), SetLastError = true, CharSet = CharSet.Auto)]
+        public static extern bool WTSEnumerateSessions(
+            IntPtr hServer,
+            [MarshalAs(UnmanagedType.U4)] int reserved,
+            [MarshalAs(UnmanagedType.U4)] int version,
+            ref IntPtr ppSessionInfo,
+            [MarshalAs(UnmanagedType.U4)] ref int pSessionInfoCount);
+
+        /// <summary>
+        /// Retrieves a list of sessions on a Remote Desktop Session Host (RD Session Host) server.
+        /// </summary>
+        /// <param name="hServer">A handle to the RD Session Host server.
+        /// You can use the <see cref="WTSOpenServer"/> or <see cref="WTSOpenServerEx"/> functions to retrieve a handle to a specific
+        /// server, or <see cref="WTS_CURRENT_SERVER_HANDLE"/> to use the RD Session Host server that hosts your application.</param>
+        /// <param name="reserved">This parameter is reserved. It must be zero.</param>
+        /// <param name="version">The version of the enumeration request. This parameter must be 1.</param>
+        /// <param name="ppSessionInfo">A pointer to <see cref="IEnumerable&lt;WTS_SESSION_INFO&gt;"/> structures that represent the retrieved
+        /// sessions. Note, that returned object doesn't know overall count of sessions, and always return true for MoveNext, use it in pair
+        /// with pSessionInfoCount parameter</param>
+        /// <param name="pSessionInfoCount">A pointer to the number of WTS_SESSION_INFO structures returned in the ppSessionInfo parameter.</param>
+        /// <returns>Returns zero if this function fails. If this function succeeds, a nonzero value is returned.</returns>
+        [DllImport(nameof(WtsApi32), SetLastError = true, CharSet = CharSet.Auto)]
+        public static extern bool WTSEnumerateSessions(
+            IntPtr hServer,
+            [MarshalAs(UnmanagedType.U4)] int reserved,
+            [MarshalAs(UnmanagedType.U4)] int version,
+            ref WtsSafeSessionInfoGuard ppSessionInfo,
```

Comment:
This will force users to use ```WtsFreeMemory``` to free this allocated memory. That is why I started with adding ```WtsSafeMemoryGuard``` and all that dangerous iterator.
These out memory is memory to free and also something that can be enumerated.
What do you think, if I rewrite this method to following:
```public static extern bool WTSEnumerateSessions(
SafeWtsServerHandle hServer, int reserved,int version,
ref WtsSafeSessionInfoGuard ppSessionInfo); // not that I don't return pSessionInfoCount
```
In this case I can use wrap ```pSessionInfoCount``` into ```WtsSafeSessionInfoGuard``` and remove it infinity iteration.